common/core/hwclayer.cpp

```diff
@@ -359,6 +359,16 @@ void HwcLayer::UpdateRenderingDamage(const HwcRect<int>& old_rect,
 }
 
 const HwcRect<int>& HwcLayer::GetLayerDamage() {
+  HwcRect<int> translated_damage = {0, 0, 0, 0};
+  if (!surface_damage_.empty() &&
```

Comment:
The issue is that validate is called after the present call. The damage needs to be updated before the call. Thus twice.
The reason for twice is that, once we call updaterenderdamage. Second, we actually calculate the translated damage and set the current_rendering_damage_

Reply:
Can we remove it from validate completely than ?

Reply:
Then it will become an assumption that the surface damage is set only after setting the source crop and display frame. Is this ok?
And updating the rendering damage in all other places becomes redundant?

Reply:
for all other place: probably so.
Assumption: Would it matter ? As overlaylayer calls this API ir-respective of source crop or damage rect change.

Reply:
Oh! I read it wrong. If we have current_rendering_damage_ as surface_damage_ in validate we would face the same issue as before. Surface damage needs to be translated (if source crop is 0,0) before assigning to current_rendering_damage_. I actually misread it as the assignment in GetLayerDamage itself i.e. `current_rendering_damage_ = translated_damage` instead of calling updaterenderingdamage.

Reply:
thx for the clean up. One more thing,, now we do these calculations even though nothing impacting the last calculations has changed. You can have an internal state tracker flag, set it when ever this needs to be re-calculated and check the flag here. If it's not set you just return immediately last calculated current_rendering_damage_